scipy/optimize/jonkervolgenant/jonkervolgenant.cpp

```diff
@@ -0,0 +1,194 @@
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+This code implements the Jonker-Volgenant algorithm for the rectangular
+assignment problem:
+
+    R. Jonker and A. Volgenant. A Shortest Augmenting Path Algorithm for
+    Dense and Sparse Linear Assignment Problems. *Computing*, 38:325-340
+    December 1987.
+
+The code is based on the pseudocode described in pages
+1685-1686 of:
+
+    DF Crouse. On implementing 2D rectangular assignment algorithms.
+    IEEE Transactions on Aerospace and Electronic Systems
+    52(4):1679-1696, August 2016
+    doi: 10.1109/TAES.2016.140952
+
+Author: PM Larsen
+*/
+
+#include <algorithm>
+#include <cmath>
+#include <vector>
+
+static int
+augmenting_path(int nc, std::vector<double>& cost, std::vector<double>& u,
+                std::vector<double>& v, std::vector<int>& path,
+                std::vector<int>& row4col,
+                std::vector<double>& shortestPathCosts, int i,
+                std::vector<bool>& SR, std::vector<bool>& SC, double* p_minVal)
+{
+    double minVal = 0;
+
+    int num_remaining = nc;
+    std::vector<int> remaining(nc);
+    // std::iota(remaining.begin(), remaining.end(), 0); //needs C++11
+    for (int it = 0; it < nc; it++) {
+        remaining[it] = it;
+    }
+
+    std::fill(SR.begin(), SR.end(), false);
+    std::fill(SC.begin(), SC.end(), false);
+    std::fill(shortestPathCosts.begin(), shortestPathCosts.end(), INFINITY);
+
+    // find shortest augmenting path
+    int sink = -1;
+    while (sink == -1) {
+
+        int index = -1;
+        double lowest = INFINITY;
+
+        for (int it = 0; it < num_remaining; it++) {
+            int j = remaining[it];
+
+            double r = minVal + cost[i * nc + j] - u[i] - v[j];
+            if (r < shortestPathCosts[j]) {
+                path[j] = i;
+                shortestPathCosts[j] = r;
+            }
+
+            if (shortestPathCosts[j] < lowest ||
+                (shortestPathCosts[j] == lowest && row4col[j] == -1)) {
+                lowest = shortestPathCosts[j];
+                index = it;
+            }
+        }
+
+        minVal = lowest;
+        int j = remaining[index];
+        if (minVal == INFINITY) { // infeasible cost matrix
+            return -1;
+        }
+
+        if (row4col[j] == -1) {
+            sink = j;
+        } else {
+            i = row4col[j];
+        }
+
+        SR[i] = true;
+        SC[j] = true;
+        remaining[index] = remaining[--num_remaining];
+        remaining.resize(num_remaining);
+    }
+
+    *p_minVal = minVal;
+    return sink;
+}
+
+static int
+solve(int nr, int nc, double* input_cost, int64_t* output_col4row)
+{
+
+    // build a non-negative cost matrix
+    std::vector<double> cost(nr * nc);
+    double minval = *std::min_element(input_cost, input_cost + nr * nc);
+    for (int i = 0; i < nr * nc; i++) {
```

Comment:
This could in principle be skipped when `minval >= 0`. Not that it would be a huge performance improvement to do so; I see about 2% of the total time being spent here in a 5000x5000 example.